### PR TITLE
Feature/writeable resources

### PR DESCRIPTION
**Pull request by annashamray**
**Changes:**
1. Make all resources writable (allow `create` and `destroy` actions)
1. Remove `datum_begin_geldigheid` and `datum_einde_geldigheid` from models with FK relation to `ZaakType`
1. Add `draft` field to all resources
1. Allow delete only for draft objects
1. Add custom action to publish resources and make them non-editable
1. Write tests

### CODE REVIEW

File: src/ztc/api/serializers/besluittype.py [hunk 1]
```diff
@@ -1,21 +1,24 @@
...
+from ...datamodel.models import BesluitType, InformatieObjectType, ZaakType
 
 
 class BesluitTypeSerializer(serializers.HyperlinkedModelSerializer):
     informatieobjecttypes = serializers.HyperlinkedRelatedField(
         view_name='informatieobjecttype-detail',
         many=True,
         lookup_field='uuid',
-        read_only=True
+        queryset=InformatieObjectType.objects.all()
```
Comment (sergei-maertens) [inline src/ztc/api/serializers/besluittype.py]: I think these need to get restricted to the same catalogue as the catalogue where the `BesluitInformatieObject` lives. Probably best to implement this as a validator?

File: src/ztc/api/serializers/besluittype.py [hunk 2]
```diff
@@ -27,6 +30,15 @@ class Meta:
...
                 'lookup_field': 'uuid',
             },
+            'begin_geldigheid': {
+                'source': 'datum_begin_geldigheid'
+            },
+            'einde_geldigheid': {
+                'source': 'datum_einde_geldigheid'
+            },
+            'draft': {
+                'read_only': True,
```
Comment (sergei-maertens) [inline src/ztc/api/serializers/besluittype.py]: ~~hrm, I'd say this depends on the current value of the field, no? If `draft` is `true` -> it's writable to set it to `false`, and after that it's read-only~~

nvm, I saw later on that it's implemented via a different action :+1:

File: src/ztc/api/serializers/roltype.py [hunk 3]
```diff
@@ -47,3 +48,13 @@ class Meta:
                 'lookup_field': 'uuid'
             }
         }
+
+    @transaction.atomic()
+    def create(self, validated_data):
```
Comment (sergei-maertens) [inline src/ztc/api/serializers/roltype.py]: ZRC has the library `drf-writable-nested` to facilitate this via a mixin :smiley:

File: src/ztc/api/serializers/zaken.py [hunk 4]
```diff
@@ -262,8 +268,23 @@ class Meta:
...
+            },
         }
 
         # expandable_fields = {
         #     'catalogus': ('ztc.api.serializers.CatalogusSerializer', {'source': 'catalogus'}),
         # }
+
+    @transaction.atomic()
+    def create(self, validated_data):
+        zaaktypen_relaties_data = validated_data.pop('zaaktypenrelaties', None)
```
Comment (sergei-maertens) [inline src/ztc/api/serializers/zaken.py]: I don't see this field defined anywhere, am I missing something?

File: src/ztc/api/tests/test_besluittype.py [hunk 5]
```diff
@@ -1,30 +1,26 @@
...
+
+from ...datamodel.models import BesluitType
+from ...datamodel.tests.factories import (
+    BesluitTypeFactory, InformatieObjectTypeFactory, ZaakTypeFactory
+)
 from .base import APITestCase
 from .utils import reverse
 
 
 class BesluitTypeAPITests(APITestCase):
```
Comment (sergei-maertens) [inline src/ztc/api/tests/test_besluittype.py]: can you add a test where you try to relate it to objects from a different catalogue, which should fail?

File: src/ztc/api/serializers/zaken.py (hunk 4 above)
Comment (annashamray) [reply src/ztc/api/serializers/zaken.py]: this is source for `gerelateerde_zaaktypen` field (which is a nested serializer):
https://github.com/VNG-Realisatie/gemma-zaaktypecatalogus/blob/906edb55968e5fb61d98c61cadc143e61a829be3/src/ztc/api/serializers/zaken.py#L128

File: src/ztc/api/serializers/besluittype.py (hunk 1 above)
Comment (annashamray) [reply src/ztc/api/serializers/besluittype.py]: validator `RelationCatalogValidator` is added to `BesluitType` and `ZaaktTpe` serializers

File: src/ztc/api/serializers/roltype.py (hunk 3 above)
Comment (annashamray) [reply src/ztc/api/serializers/roltype.py]: replace manual `create`with `NestedCreateMixin` from drf-writable-nested for `RolType` and `ZaakType` serializers

File: src/ztc/api/tests/test_besluittype.py (hunk 5 above)
Comment (annashamray) [reply src/ztc/api/tests/test_besluittype.py]: additional tests for `besluittype` and `zaaktype` creation are added

**Pull request merged**